st2common/st2common/ssh/parallel_ssh.py

```diff
@@ -0,0 +1,212 @@
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+import traceback
+
+import eventlet
+
+from st2common import log as logging
+from st2common.ssh.paramiko_ssh import ParamikoSSHClient
+import st2common.util.jsonify as jsonify
+
+LOG = logging.getLogger(__name__)
+
+
+class ParallelSSHClient(object):
+    KEYS_TO_TRANSFORM = ['stdout', 'stderr']
+    CONNECT_ERROR = 'Cannot connect to host.'
+
+    def __init__(self, hosts, user=None, password=None, pkey=None, port=22, concurrency=10,
+                 raise_on_error=False, connect=True):
+        self._ssh_user = user
+        self._ssh_key = pkey
+        self._ssh_password = password
+        self._hosts = hosts
+        self._ssh_port = port
+
+        if not hosts:
+            raise Exception('Need an non-empty list of hosts to talk to.')
+
+        self._pool = eventlet.GreenPool(concurrency)
+        self._hosts_client = {}
+        self._bad_hosts = {}
+        self._scan_interval = 0.1
+        if connect:
+            self.connect(raise_on_error=raise_on_error)
+
+    def connect(self, raise_on_error=False):
+        for host in self._hosts:
+            LOG.info('Connecting to host: %s as user: %s password: %s and key: %s', host,
+                     self._ssh_user, self._ssh_password, self._ssh_key)
+            client = ParamikoSSHClient(host, username=self._ssh_user, password=self._ssh_password,
+                                       key=self._ssh_key, port=self._ssh_port)
+            try:
+                client.connect()
+            except:
+                error = 'Failed connecting to host %s.' % host
+                LOG.exception(error)
+                _, ex, tb = sys.exc_info()
+                if raise_on_error:
+                    raise
+                self._bad_hosts[host] = {
+                    'traceback': tb,
+                    'error': ' '.join([self.CONNECT_ERROR, str(ex)])
+                }
+            else:
+                self._hosts_client[host] = client
+
+    def run(self, cmd, timeout=None, cwd=None):
+        # Note that doing a chdir using sftp client in ssh_client doesn't really
+        # set the session working directory. So we have to do this hack.
+        if cwd:
+            cmd = 'cd %s && %s' % (cwd, cmd)
```

Comment:
`cwd` is provided by the user, right?

If so, we should quote / escape it to avoid injections (you might argue that user can already run commands / script on the host so they already have all the privileges, but this might not be the case if user is only allowed to run a particular script).